core/src/main/java/com/linecorp/armeria/common/QueryStringEncoder.java

```diff
@@ -0,0 +1,241 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright 2012 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.common;
+
+import java.util.Map.Entry;
+
+import com.linecorp.armeria.internal.TemporaryThreadLocals;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufUtil;
+import io.netty.util.internal.StringUtil;
+
+final class QueryStringEncoder {
+
+    // Forked from netty-4.1.43.
+    // https://github.com/netty/netty/blob/bd8cea644a07890f5bada18ddff0a849b58cd861/codec-http/src/main/java/io/netty/handler/codec/http/QueryStringEncoder.java
+
+    private static final byte WRITE_UTF_UNKNOWN = (byte) '?';
+    private static final char[] CHAR_MAP = "0123456789ABCDEF".toCharArray();
+    private static final boolean[] SAFE_OCTETS =
+            createSafeOctets("-_.*abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789");
+
+    private static boolean[] createSafeOctets(String safeChars) {
+        int maxChar = -1;
+        for (int i = 0; i < safeChars.length(); i++) {
+            maxChar = Math.max(safeChars.charAt(i), maxChar);
+        }
+        final boolean[] octets = new boolean[maxChar + 1];
+        for (int i = 0; i < safeChars.length(); i++) {
+            octets[safeChars.charAt(i)] = true;
+        }
+        return octets;
+    }
+
+    static void encodeParams(TemporaryThreadLocals tempThreadLocals,
+                             StringBuilder buf, QueryParamGetters params) {
+        for (Entry<String, String> e : params) {
+            final String name = e.getKey();
+            if (isSafeOctetsOnly(name)) {
+                buf.append(name);
+            } else {
+                encodeUtf8Component(tempThreadLocals, buf, name);
+            }
+            buf.append('=');
+
+            final String value = e.getValue();
+            if (isSafeOctetsOnly(value)) {
+                buf.append(value);
+            } else {
+                encodeUtf8Component(tempThreadLocals, buf, value);
+            }
+            buf.append('&');
+        }
+
+        buf.setLength(buf.length() - 1);
+    }
+
+    private static boolean isSafeOctetsOnly(String s) {
```

Comment:
Compared to Guava, we seem to lose a lot by not returning an index here and using it in `encodeUtf8Component` to skip checking initial characters again.

Reply:
Fixed